### PR TITLE
Implement ActionInterface for /customer/section/load

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)
This PR is example for **How to replace Abstract Action class with ActionInterface**

That specific change is important due to huge performance improvement:
image

### Fixed Issues (if relevant)
N/A

### Manual testing scenarios (*)
N/A

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Customer/Controller/Section/Load.php [hunk 1]
```diff
@@ -3,85 +3,83 @@
...
         SectionPoolInterface $sectionPool,
-        \Magento\Framework\Escaper $escaper = null
+        Escaper $escaper
     ) {
-        parent::__construct($context);
         $this->resultJsonFactory = $resultJsonFactory;
-        $this->sectionIdentifier = $sectionIdentifier;
         $this->sectionPool = $sectionPool;
-        $this->escaper = $escaper ?: $this->_objectManager->get(\Magento\Framework\Escaper::class);
+        $this->escaper = $escaper ?? ObjectManager::getInstance()->get(Escaper::class);
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Controller/Section/Load.php]: We don’t need to have this fallback

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Customer/Controller/Section/Load.php (hunk 1 above)
Comment (lbajsarowicz) [reply app/code/Magento/Customer/Controller/Section/Load.php]: Aww... That is actually a huge mistake. Thank you!

**Review by ihor-sviziev — APPROVED** [review]
Failing functional test are not related to changes from this PR

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7103](https://jira.corp.magento.com/browse/ENGCOM-7103) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Comment by lbajsarowicz** [comment]
@magento run all tests
False negative on Static Tests

File: app/code/Magento/Customer/Controller/Section/Load.php [hunk 2]
```diff
@@ -3,85 +3,83 @@
...
  */
+declare(strict_types=1);
+
 namespace Magento\Customer\Controller\Section;
 
-use Magento\Framework\App\Action\HttpGetActionInterface as HttpGetActionInterface;
-use Magento\Customer\CustomerData\Section\Identifier;
 use Magento\Customer\CustomerData\SectionPoolInterface;
-use Magento\Framework\App\Action\Context;
+use Magento\Framework\App\Action\HttpGetActionInterface as HttpGetActionInterface;
```
Comment (markshust) [inline app/code/Magento/Customer/Controller/Section/Load.php]: redundant `HttpGetActionInterface` alias

Comment (lbajsarowicz) [reply app/code/Magento/Customer/Controller/Section/Load.php]: Right, thank you!

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
Hi, @lbajsarowicz, resolve conflict. Thanks.

**Comment by lbajsarowicz** [comment]
Of course, I can.

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by ihor-sviziev** [comment]
@engcom-Echo approved from my side, please take it

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7103](https://jira.corp.magento.com/browse/ENGCOM-7103) has been created to process this Pull Request

**Review by engcom-Delta — CHANGES_REQUESTED** [review]
Hi @lbajsarowicz , there is an issue with PR changes: refresh home page on second store view returns customer to default sore view

**Steps to reproduce:**

- Generate testing categories and products
- Create second store view

image

- Go to Home page
- Change store view to Second Store View
- Refresh page

:heavy_check_mark: **Expected result:**
Home page of Second Store View is shown
#27296Main


:x: **Actual result:**
Home page of Default Store View is shown
#27296PR

Could you take a look?

**Comment by ihor-sviziev** [comment]
Hi @lbajsarowicz ,
Will you be able to update your PR?

**Comment by lbajsarowicz** [comment]
@ihor-sviziev Yes, but I've just finished very, very difficult week for me and I need to get rest a few hours.

**Comment by lbajsarowicz** [comment]
I remember about this PR :-)

**Comment by lbajsarowicz** [comment]
@engcom-Delta Thank you very much! You detected leftover in our DI after AbstractActions and Actions! You are awesome. The issue is finally fixed and you can test latest version of branch!

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @lbajsarowicz,
Everything looks good, but static tests should started failing. Please fix

**Comment by lbajsarowicz** [comment]
I'm writing MFTF test to cover the reported issue.

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7103](https://jira.corp.magento.com/browse/ENGCOM-7103) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
Was checked customer section load without errors and works as before
Peek 2020-04-22 13-42

Was checked case from [comment](https://github.com/magento/magento2/pull/27296#pullrequestreview-381138334)

File: app/code/Magento/Customer/Controller/Section/Load.php [hunk 3]
```diff
@@ -3,85 +3,83 @@
...
+use Magento\Framework\App\RequestInterface;
+use Magento\Framework\Controller\Result\Json as JsonResult;
 use Magento\Framework\Controller\Result\JsonFactory;
+use Magento\Framework\Escaper;
 
 /**
- * Customer section controller
+ * Endpoint `customer/section/load` responsible for reloading sections of Customer's Local Storage
  */
-class Load extends \Magento\Framework\App\Action\Action implements HttpGetActionInterface
```
Comment (lenaorobei) [inline app/code/Magento/Customer/Controller/Section/Load.php:17]: Please revert this line for BC.

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Please see comment.

**Comment by okorshenko** [comment]
Hi @lbajsarowicz 
That's very interesting PR and performance measurements. We are actively working with architects and performance team to find the best option (considering inheritance issue, BiC and the way to address it, performance improvements)

**Comment by lbajsarowicz** [comment]
@okorshenko The comparison was done between `2.4-develop` without ActionInterface changes and after ActionInterface Changes + removal of inheritance. Measured today, that change just between having inheritance and having just composition is not just as big. So even if we use inheritance, we are going to lose 10 - 15% of performance at most (7% on my local lab)

**Comment by ihor-sviziev** [comment]
FYI this PR was moved to "oh hold" due to internal discussions, reason the same as there https://github.com/magento/magento2/pull/27538#issuecomment-624891873

**Review by lenaorobei — APPROVED** [review]
During the migration we faced with the issue that may occur for 3rd-party developers who use observers which subscribe to those controllers and are trying to call methods, which were eliminated.
Despite the fact that controllers are not the part of our public API, we pass controller to the event. Its structure is the part of public API.

### Taking that it into account here is what was decided:

- we recommend using interface approach starting 2.4.0 for new code only
- we will not migrate existing core controllers until 2.5.0
- we agreed to accept already open PRs as refactoring to 2.4.1 but will require to return extends in order to keep BC

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-7103](https://jira.corp.magento.com/browse/ENGCOM-7103) has been created to process this Pull Request

**Comment by lbajsarowicz** [comment]
If anyone else wants to pick this, feel free to take my code and continue.
I won't be available to handle that in next 2 - 3 weeks.

**Pull request closed without merging**